src/librustc_driver/lib.rs

```diff
@@ -201,45 +203,137 @@ mod no_llvm_metadata_loader {
+            MetadataOnlyTransCrate
+        }
+    }
+
+    impl TransCrate for MetadataOnlyTransCrate {
+        type MetadataLoader = ::no_llvm_metadata_loader::NoLlvmMetadataLoader;
+        type OngoingCrateTranslation = OngoingCrateTranslation;
+        type TranslatedCrate = TranslatedCrate;
+
+        fn metadata_loader() -> Box<MetadataLoader> {
+            box ::no_llvm_metadata_loader::NoLlvmMetadataLoader
+        }
+
+        fn provide(_providers: &mut Providers) {}
+
+        fn trans_crate<'a, 'tcx>(
+            tcx: TyCtxt<'a, 'tcx, 'tcx>,
+            analysis: CrateAnalysis,
+            incr_hashes_map: IncrementalHashesMap,
+            _output_filenames: &OutputFilenames
+        ) -> Self::OngoingCrateTranslation {
+            let link_meta = build_link_meta(&incr_hashes_map);
+            let exported_symbols = find_exported_symbols(tcx, &analysis.reachable);
+            let (metadata, _hashes) = tcx.encode_metadata(&link_meta, &exported_symbols);
+
+            OngoingCrateTranslation {
+                metadata: metadata,
+                metadata_version: tcx.metadata_encoding_version().to_vec(),
+                crate_name: tcx.crate_name(LOCAL_CRATE),
+            }
+        }
+
+        fn join_trans(
+            trans: Self::OngoingCrateTranslation,
+            _sess: &Session,
+            _dep_graph: &DepGraph,
+        ) -> Self::TranslatedCrate {
+            TranslatedCrate(trans)
+        }
+
+        fn link_binary(sess: &Session,
+                       trans: &Self::TranslatedCrate,
+                       outputs: &OutputFilenames) {
+            for &crate_type in sess.opts.crate_types.iter() {
+                if crate_type != CrateType::CrateTypeRlib &&
+                   crate_type != CrateType::CrateTypeDylib {
+                    continue;
+                }
+                let output_name =
+                    out_filename(sess, crate_type, &outputs, &trans.0.crate_name.as_str());
+                let mut compressed = trans.0.metadata_version.clone();
+                let metadata = if crate_type == CrateType::CrateTypeDylib {
+                    DeflateEncoder::new(&mut compressed, Compression::Fast)
+                        .write_all(&trans.0.metadata.raw_data).unwrap();
+                    &compressed
+                } else {
+                    &trans.0.metadata.raw_data
+                };
+                let mut builder = Builder::new(File::create(&output_name).unwrap());
+                let header = Header::new(
+                    "rust.metadata.bin".to_string(),
+                    metadata.len() as u64
+                );
+                builder
+                    .append(&header, Cursor::new(metadata))
+                    .unwrap();
+            }
+        }
+
+        fn dump_incremental_data(_trans: &Self::TranslatedCrate) {}
+    }
+}
 
-    use self::back::write::OngoingCrateTranslation;
+#[cfg(not(feature="llvm"))]
+mod rustc_trans {
```

Comment:
Can you move all of these items to the `TransCrate` trait?

Reply:
No, because most of them are llvm specific, eg they would still use llvm even if trans is non llvm. (enable_llvm_debug, print_passes)